mozilla/bedrock
# Postgres-to-sqlite DB exporter script: redux

This changeset is focused on the DB export script, which takes the state of Bedrock from the Postgres DB and exports it as a Sqlite DB, but excludes any sensitive data or PII.

This script focuses purely on DB generation. A separate PR will slot it in to our infra and upload the exported DB to our regular cloud bucket, ready for use elsewhere in our local dev, test and demo deployments. (Issue: https://github.com/mozilla/bedrock/issues/14660)

The previous cut of this script worked, but operated in a subtractive/denying approach. As much as is practically possible, this new version is additive and allowlist-based, which vastly reduces the scope for accidental data leakage in the future if new DB tables are added, for example.

Note that for integrity reasons, we do have to _temporarily_ export and load `auth.User` and `wagtailcore.Revision` data, but we take pains to null out foreign keys to them, then delete all records from the relevant tables, and then double-check the tables are empty. 

We also delete the output database if a problem is encountered along the way, so if the export process fails for some reason there will be no half-cleaned exported database available.

The DB tables for CMS Images and Documents are exoprted, but the corresponding files are not - that will come in a separate, on-demand command to avoid wasted time, CPU and bandwidth: https://github.com/mozilla/bedrock/issues/14659


Resolves #14937
## Reviewing

**Some parts of this code (the sqlite/SQL and the bash loops) were written with the assistance of a GPT -- but then sense checked by me, of course**

I urge you to get a good cup of coffee and settle in to review this - please be sceptical and picky, and please do check you feel you definitely understand what it's doing at each stage (e.g. why certain tables are [not] included, etc)

## Testing

* Set up a CMS user if you don't have one already: [docs are here](https://bedrock.readthedocs.io/en/latest/cms.html#accessing-the-cms) and you can use SSO or non-SSO as you prefer.

* Set up a local postgres DB from which to export:
1. `docker compose up db` will start a postgres DB on your local machine
2. Create a `bedrock` DB in there:
```
docker compose exec db bash
psql -U postgres
create database bedrock;
\q   # to quit psql
exit  # to leave the Docker container
```
3. Fill the postgres DB from the bedrock sqlite DB you already have, using a script that is not in this PR: `DATABASE_URL="postgres://postgres:postgres@127.0.0.1:5432/bedrock" ./bin/fill-empty-postgres-database.sh`
5. In your .env set `DATABASE_URL="postgres://postgres:postgres@127.0.0.1:5432/bedrock"`
6. Restart your runserver and confirm you can still browse the site. 
7. Optional: in the CMS, add a child page to the homepage and publish it. View it live to confirm it's OK

* Export the postgres DB to a new sqlite file:
1. DATABASE_URL="postgres://postgres:postgres@127.0.0.1:5432/bedrock" ./bin/export-db-to-sqlite.sh /Users/steve/Code/bedrock/data/exported.db

(or run it in debug mode `1. DATABASE_URL="postgres://postgres:postgres@127.0.0.1:5432/bedrock" bash -x ./bin/export-db-to-sqlite.sh /Users/steve/Code/bedrock/data/exported.db` to see all the extra context/info.

2. In your .env set `DATABASE_URL="sqlite://bedrock/data/exported.db"`
3. Restart your runserver and browse around - everything should be fine, and you should also be able to visit the same page you just created in the postgres-backed version of your local site.
